pkg/apis/networking/validation/validation.go

```diff
@@ -597,3 +598,78 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(spec.NodeSelector, fldPath.Child("nodeSelector"))...)
+
+	maxIPv4PerNodeMaskSize := int32(32)
+	maxIPv6PerNodeMaskSize := int32(128)
+
+	hasIPv4 := spec.IPv4 != nil
+	hasIPv6 := spec.IPv6 != nil
+
+	// Validate if CIDR is configured for at least one IP Family(IPv4/IPv6).
+	if !hasIPv4 && !hasIPv6 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv4", "ipv6"), *spec, "either IPv4 or IPv6 CIDR must be configured, both cannot be empty"))
+	}
+
+	// Validate configured IPv4 CIDR and PerNodeMaskSize.
+	if hasIPv4 {
+		if !netutils.IsIPv4CIDRString(spec.IPv4.CIDR) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv4", "cidr"), *spec.IPv4, "provided CIDR is not a valid IPv4 CIDR"))
+		}
+		if spec.IPv4.PerNodeMaskSize < 0 || spec.IPv4.PerNodeMaskSize > maxIPv4PerNodeMaskSize {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv4", "perNodeMaskSize"), *spec.IPv4, "provided IPv4 PerNodeMaskSize is invalid, Valid values: [0,32]"))
+		}
+	}
+
+	// Validate configured IPv6 CIDR and PerNodeMaskSize.
+	if hasIPv6 {
+		if !netutils.IsIPv6CIDRString(spec.IPv6.CIDR) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv6", "cidr"), *spec.IPv6, "provided CIDR is not a valid IPv6 CIDR"))
+		}
+		if spec.IPv6.PerNodeMaskSize < 0 || spec.IPv6.PerNodeMaskSize > maxIPv6PerNodeMaskSize {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv6", "perNodeMaskSize"), *spec.IPv6, "provided IPv6 PerNodeMaskSize is invalid, Valid values: [0,128]"))
+		}
+	}
+
+	// Validate PerNodeMaskSize if both IPv4 and IPv6 CIDRs are configured.
+	// 32 - spec.IPv4.PerNodeMaskSize == 128 - spec.IPv6.PerNodeMaskSize must be satisfied.
+	if hasIPv4 && hasIPv6 {
+		if maxIPv4PerNodeMaskSize-spec.IPv4.PerNodeMaskSize != maxIPv6PerNodeMaskSize-spec.IPv6.PerNodeMaskSize {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv4", "ipv6", "perNodeMaskSize"), *spec, "provided IPv4 and IPv6 PerNodeMaskSizes are invalid, Valid Values: 32 - IPv4.PerNodeMaskSize == 128 - IPv6.PerNodeMaskSize"))
+		}
+	}
+	return allErrs
+}
+
+// ValidateClusterCIDRConfigUpdate tests if an update to a ClusterCIDRConfig is valid.
+func ValidateClusterCIDRConfigUpdate(update, old *networking.ClusterCIDRConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, apivalidation.ValidateObjectMetaUpdate(&update.ObjectMeta, &old.ObjectMeta, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, validateClusterCIDRConfigUpdateSpec(&update.Spec, &old.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+func validateClusterCIDRConfigUpdateSpec(update, old *networking.ClusterCIDRConfigSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// ClusterCIDRConfig Spec is immutable, no updates are allowed.
+	if cmp.Diff(update, old, cmp.Options{}) != "" {
```

Comment:
I remember some discussion not recommending using this in one review but I can't find it, I could only find this https://github.com/kubernetes/kubernetes/issues/55722  ... I think that will be better to copy the same logic used in other APIs

https://github.com/kubernetes/kubernetes/blob/b396cd6f8ffbe07bf6a23b259b3063dec496dfe9/pkg/api/validation/validation.go#L1554-L1565

Reply:
Updated it in https://github.com/kubernetes/kubernetes/pull/108290/commits/4f46ccd81c234bbbf83983c1edb471201d85de4c#diff-b955dad08cd849f27ea132e779f2f62e5edf03ee55a58424c7ff5e218ce6d487L670

Moved away from using cmp.Diff